actions/views/channel.js

```diff
@@ -20,7 +27,82 @@ export function checkAndSetMobileView() {
 
 export function goToLastViewedChannel() {
     return async (dispatch, getState) => {
-        const lastViewedChannel = getChannelByName(getState(), getLastViewedChannelName(getState()));
-        dispatch(selectChannel(lastViewedChannel.id));
+        const state = getState();
+        const lastViewedChannel = getChannelByName(state, getLastViewedChannelName(state));
+        return dispatch(switchToChannel(lastViewedChannel));
+    };
+}
+
+export function switchToChannelById(channelId) {
+    return async (dispatch, getState) => {
+        const state = getState();
+        const channel = getChannel(state, channelId);
+        return dispatch(switchToChannel(channel));
+    };
+}
+
+export function switchToChannel(channel) {
+    return async (dispatch, getState) => {
+        const state = getState();
+        const teamUrl = getCurrentRelativeTeamUrl(state);
+
+        if (channel.fake || channel.userId) {
+            const username = channel.userId ? channel.name : channel.display_name;
+            const user = getUserByUsername(state, username);
+            if (!user) {
+                return {error: true};
+            }
+
+            const direct = await dispatch(openDirectChannelToUserId(user.id));
+            if (direct.error) {
+                return {error: true};
+            }
+            browserHistory.push(`${teamUrl}/messages/@${channel.name}`);
+        } else {
+            browserHistory.push(`${teamUrl}/channels/${channel.name}`);
+        }
+
+        return {data: true};
+    };
+}
+
+export function openDirectChannelToUserId(userId) {
+    return async (dispatch, getState) => {
+        const state = getState();
+        const currentUserId = getCurrentUserId(state);
+        const channelName = getDirectChannelName(currentUserId, userId);
+        const channel = getChannelByName(state, channelName);
+
+        if (channel) {
```

Comment:
Minor style thing, this if statement could be negated with an early return to avoid nesting so much